File: app/src/main/java/com/example/bloodbank/ui/fragment/homeCycle/NotificationFragment.java
package com.example.bloodbank.ui.fragment.homeCycle;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.bloodbank.R;
import com.example.bloodbank.adapter.NotificationAdapter;
import com.example.bloodbank.data.model.notifications.Notification;
import com.example.bloodbank.data.model.notifications.NotificationData;
import com.example.bloodbank.ui.fragment.BaseFragment;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.example.bloodbank.data.api.ApiClient.getClient;
import static com.example.bloodbank.data.local.SharedPreferencesManger.API_TOKEN;
import static com.example.bloodbank.data.local.SharedPreferencesManger.LoadData;
import static com.example.bloodbank.helper.HelperMethods.dismissProgressDialog;
import static com.example.bloodbank.helper.HelperMethods.showProgressDialog;

public class NotificationFragment extends BaseFragment {
    @BindView(R.id.fragment_notification_rv)
    RecyclerView fragmentNotificationRv;

    private List<NotificationData> notificationData;
    private NotificationAdapter notificationAdapter;
    private String TAG = "notification";
    private String apiToken;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        setupAcvtivity();
        View view = inflater.inflate(R.layout.fragment_notification, container, false);
        ButterKnife.bind(this, view);

        apiToken = LoadData(getActivity(), API_TOKEN);
//        apiToken = "W4mx3VMIWetLcvEcyF554CfxjZHwdtQldbdlCl2XAaBTDIpNjKO1f7CHuwKl";

        getNotification(1);

        initRecyc();
        return view;
    }

    private void initRecyc() {
        notificationData = new ArrayList<>();
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getActivity());
        notificationAdapter = new NotificationAdapter(getActivity(), notificationData);
        fragmentNotificationRv.setLayoutManager(linearLayoutManager);
        fragmentNotificationRv.setAdapter(notificationAdapter);
    }


    private void getNotification(int page) {
        showProgressDialog(getActivity(), getString(R.string.please_wait));
        getClient().getNotifications(apiToken, page).enqueue(new Callback<Notification>() {
            @Override
            public void onResponse(Call<Notification> call, Response<Notification> response) {
                dismissProgressDialog();
                try {
                    if (response.body().getStatus() == 1) {
                        notificationData.addAll(response.body().getData().getData());
                        notificationAdapter.notifyDataSetChanged();
                    } else {
                        Log.i(TAG, response.body().getMsg());
                    }
                } catch (
                        Exception e) {
                    Log.i(TAG, e.getMessage());
                }
            }

            @Override
            public void onFailure(Call<Notification> call, Throwable t) {

            }
        });
    }

    @Override
    public void onBack() {
        super.onBack();
    }
}

File: app/src/main/java/com/example/bloodbank/ui/fragment/homeCycle/notifications/NotificationFragment.kt
package com.example.bloodbank.ui.fragment.homeCycle.notifications

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.bloodbank.R
import com.example.bloodbank.data.model.notifications.NotificationData
import com.example.bloodbank.databinding.FragmentNotificationBinding
import com.example.bloodbank.extensions.createToast
import com.example.bloodbank.extensions.inflateWithBinding
import com.example.bloodbank.ui.adapter.NotificationAdapter
import com.example.bloodbank.ui.fragment.BaseFragment
import com.example.bloodbank.ui.views.LoadingDialog
import com.example.bloodbank.utils.API_TOKEN

class NotificationFragment : BaseFragment(), NotificationContract.View {

    lateinit var binding: FragmentNotificationBinding
    private var notificationData: MutableList<NotificationData> = mutableListOf()
    private lateinit var notificationAdapter: NotificationAdapter
    private lateinit var apiToken: String

    lateinit var notificationPresenter: NotificationPresenter
    lateinit var loadingDialog: LoadingDialog

    lateinit var linearLayoutManager: LinearLayoutManager
    private var pageNo = 1
    private var previousTotal: Int = 0
    private var visibleThreshold: Int = 5
    var loading: Boolean = true

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        setupActivity()
        binding = inflater.inflateWithBinding(R.layout.fragment_notification, container)
        notificationPresenter = NotificationPresenter(this)
        init()
        notificationPresenter.requestDataFromServer(apiToken)
        setListener()
        return binding.root
    }

    private fun init() {
        apiToken = API_TOKEN!!
        loadingDialog = LoadingDialog(activity)

        notificationAdapter = NotificationAdapter(notificationData)
        linearLayoutManager = LinearLayoutManager(activity)
        binding.rvNotificationList.layoutManager = linearLayoutManager
        binding.rvNotificationList.adapter = notificationAdapter
    }

    private fun setListener() {

        binding.rvNotificationList.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)

                val visibleItemCount = binding.rvNotificationList.childCount
                val totalItemCount = linearLayoutManager.itemCount
                val firstVisibleItem = linearLayoutManager.findFirstVisibleItemPosition()


                if (loading) {
                    if (totalItemCount > previousTotal) {
                        loading = false
                        previousTotal = totalItemCount
                    }
                }
                if (!loading && totalItemCount - visibleItemCount
                        <= firstVisibleItem + visibleThreshold) {
                    notificationPresenter.getMoreData(apiToken, pageNo)
                    loading = true
                }
            }
        })
    }

    override fun showProgress() {
        loadingDialog.startLoadingDialog(getString(R.string.please_wait))
    }

    override fun hideProgress() {
        loadingDialog.dismissDialog()
    }

    override fun onResponseFailure(message: String) {
        requireActivity().createToast(message)
    }

    override fun setDataToRecyclerView(notificationList: MutableList<NotificationData>) {
        notificationData.addAll(notificationList)
        notificationAdapter.notifyDataSetChanged()
        pageNo++
    }
}